### ISSUE TITLE
OCaml 4.08beta2 fails to link on NetBSD/arm64

### ISSUE THREAD

**Issue by jaapb** [ocaml/ocaml#8567]
When compiling OCaml 4.08beta2 on a Pinebook running NetBSD (arm64 architecture), it fails with the following error message:

```
gmake[1]: Entering directory '/usr/pkgsrc/wip/ocaml/work/ocaml-4.08.0+beta2'
boot/ocamlrun ./ocamlopt -g -nostdlib -I stdlib -I otherlibs/dynlink  -o ocamlc.opt compilerlibs/ocamlcommon.cmxa compilerlibs/ocamlbytecomp.cmxa driver/main.cmx -cclib "-lm  -lpthread "
ld: stdlib/libasmrun.a(arm64.o): in function `caml_raise_exn':
(.text+0x324): relocation truncated to fit: R_AARCH64_LDST64_ABS_LO12_NC against symbol `caml_backtrace_active' defined in .bss section in stdlib/libasmrun.a(backtrace_n.o)
ld: (.text+0x324): warning: one possible cause of this error is that the symbol is being referenced in the indicated code as if it had a larger alignment than was declared where it was defined
ld: stdlib/libasmrun.a(arm64.o): in function `caml_raise_exception':
(.text+0x374): relocation truncated to fit: R_AARCH64_LDST64_ABS_LO12_NC against symbol `caml_backtrace_active' defined in .bss section in stdlib/libasmrun.a(backtrace_n.o)
ld: (.text+0x374): warning: one possible cause of this error is that the symbol is being referenced in the indicated code as if it had a larger alignment than was declared where it was defined
File "caml_startup", line 1:
Error: Error during linking
```

I'm using the package available at [https://wip.pkgsrc.org/cgi-bin/gitweb.cgi?p=pkgsrc-wip.git;a=tree;f=ocaml](url) which adds a few local patches, but nothing that I can see would cause this. It does patch configure to recognise NetBSD/aarch64 as an arm64 architecture that has dynamic linking.

Ocaml 4.07 compiles fine with a similar patch.

I've seen issue #6283 - but that seems to occur in a different part in the tree, and I've checked that the changes applied in that issue are still there.

**Comment by gasche** [comment]
This may be related to the recent alignment-fix #8557 , which was merged in trunk but not, I think, in the 4.08 release branch (and wouldn't be in beta2 anyway). Would you mind trying to build the `trunk` on your machine and report if that works?

**Comment by mshinwell** [comment]
Is this happening because `caml_backtrace_active` is defined as a 32-bit quantity in `runtime/backtrace.c`, but being loaded using an instruction whose relocation expects a 64-bit-aligned address, in `runtime/arm64.S` (e.g. line 419)?

**Comment by jaapb** [comment]
I can confirm that trunk, with the same patches and configure options, builds for me.

**Comment by xavierleroy** [comment]
@mshinwell is right, the wrong instructions are used to access `caml_backtrace_active`.  The problem should show up as a link-time error when the linker puts `caml_backtrace_active` at an address = 4 mod 8.  It may also result in backtraces being stashed while `caml_backtrace_active` is actually 0.

I have been unable to force the Linux linker to do the 4 mod 8 placement, but apparently the NetBSD one can do it. The fix #8557 changes the placement of global data, so it can hide the problem, but is not a proper fix.

**Comment by xavierleroy** [comment]
Proposed fix: #8569
